### PR TITLE
PSReadLine integration (WIP)

### PR DESCRIPTION
**Pull request by SeeminglyScience**
Need to resolve a significant amount of merge conflicts, but I wanted to get this up so folks know it's close.  Tests probably need some (or a lot of) fixing as well.

The appveyor build is unlikely to work until I resolve the merge conflicts due to changes to the startup script. Also requires the PSReadLine build from this PR lzybkr/PSReadLine#626

Highlights:

* PSReadLine integration
* Tons of stability fixes, especially in the debugger
* `$Host.EnterNestedPrompt()` support

I'd suggest holding off on formally reviewing until the merge conflicts are resolved.  After that's done I'll also update this description with significantly more detail about the changes.

Anyone curious about the changes required for PSRL should check out specifically the classes PowerShellContext, InvocationEventQueue, PromptNest, and PSReadLinePromptContext.

### CODE REVIEW

File: src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        private const string ReadLineScript = @"
+            [System.Diagnostics.DebuggerHidden()]
+            [System.Diagnostics.DebuggerStepThrough()]
+            param()
+            return [Microsoft.PowerShell.PSConsoleReadLine]::ReadLine(
+                $Host.Runspace,
+                $ExecutionContext,
+                $args[0])";
+
+        // private const string ReadLineScript = @"
```
Comment (bergmeister) [inline src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs]: remove or re-consider commented out code

Comment (SeeminglyScience) [reply src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs]: Nice catch! Must have missed that one on my sweep :)

Comment (bergmeister) [reply src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs]: No worries, as you said, the full review will only come after the conflict resolution and especial when it is not WIP any more.

File: src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs [hunk 2]
```diff
@@ -779,8 +804,7 @@ protected void Stop()
                         i));
             }
 
-            await requestContext.SendResult(
-                new StackTraceResponseBody
+            await requestContext.SendResult( new StackTraceResponseBody
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices.Protocol/Server/DebugAdapter.cs]: Mistake, must have done a vim line join and didn't notice.

File: src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs [hunk 3]
```diff
@@ -1463,6 +1463,15 @@ private static FileChange GetFileChangeDetails(Range changeRange, string insertS
             catch (TaskCanceledException)
             {
                 // If the task is cancelled, exit directly
+                foreach (var script in filesToAnalyze)
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices.Protocol/Server/LanguageServer.cs]: Comment needs to be updated.

File: src/PowerShellEditorServices/Debugging/DebugService.cs [hunk 4]
```diff
@@ -851,43 +935,54 @@ private bool AddToAutoVariables(PSObject psvariable, string scope)
...
 
-            for (int i = 0; i < callStackFrames.Length; i++)
-            {
-                VariableContainerDetails autoVariables =
-                    new VariableContainerDetails(
-                        this.nextVariableId++,
-                        VariableContainerDetails.AutoVariablesName);
+            // If access to stackFrameDetails isn't controlled there is a race condition where
+            // the array isn't finished populating before
+            // await this.stackFramesHandle.WaitAsync();
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Debugging/DebugService.cs]: More commented code that should be removed or uncommented.

File: src/PowerShellEditorServices/Debugging/DebugService.cs [hunk 5]
```diff
@@ -851,43 +935,54 @@ private bool AddToAutoVariables(PSObject psvariable, string scope)
...
+                        !string.Equals(stackFrameScriptPath, StackFrameDetails.NoFileScriptPath))
+                    {
+                        this.stackFrameDetails[i].ScriptPath =
+                            this.remoteFileManager.GetMappedPath(
+                                stackFrameScriptPath,
+                                this.powerShellContext.CurrentRunspace);
+                    }
                 }
-            }
+            // }
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Debugging/DebugService.cs]: More commented code that should be removed or uncommented.

File: src/PowerShellEditorServices/Language/CommandHelpers.cs [hunk 6]
```diff
@@ -52,6 +52,21 @@ public class CommandHelpers
                 return null;
             }
 
+            // Keeping this commented out for now.  It would be faster, but it doesn't automatically
+            // import modules. This may actually be preferred, but it's a big change that needs to
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Language/CommandHelpers.cs]: This comment should be removed and replaced with an issue.

File: src/PowerShellEditorServices/Session/Host/EditorServicesPSHostUserInterface.cs [hunk 7]
```diff
@@ -850,45 +880,50 @@ private void PowerShellContext_DebuggerResumed(object sender, System.Management.
...
                     this.WriteOutput(string.Empty);
-                    this.ShowCommandPrompt();
                 }
             }
+            else if (
+                eventArgs.ExecutionOptions.WriteErrorsToHost &&
+                (eventArgs.ExecutionStatus == ExecutionStatus.Failed ||
+                    eventArgs.HadErrors))
+            {
+                // this.CancelCommandPrompt();
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/Host/EditorServicesPSHostUserInterface.cs]: Commented code needs to be removed.

File: src/PowerShellEditorServices/Session/InvocationEventQueue.cs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            // We need to create the PowerShell object in the same thread so we can get a nested
+            // PowerShell.  Without changes to PSReadLine directly, this is the only way to achieve
+            // that consistently.  The alternative is to make the subscriber a script block and have
+            // that create and process the PowerShell object, but that puts us in a different
+            // SessionState and is a lot slower.
+
+            // This should be safe as PSReadline should be waiting for pipeline input due to the
+            // OnIdle event sent along with it.
+            typeof(PSEventSubscriber)
+                .GetProperty(
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/InvocationEventQueue.cs]: PropertyInfo needs to be stored statically.

File: src/PowerShellEditorServices/Session/InvocationEventQueue.cs [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+            internal void Invoke(PowerShell pwsh)
+            {
+                try
+                {
+                    _invocationAction(pwsh);
+
+                    // Ensure the result is set in another thread otherwise the caller
+                    // may take over the pipeline thread.
+                    System.Threading.Tasks.Task.Run(() => SetResult(true));
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/InvocationEventQueue.cs]: Replace fully qualified type name with using statement.

File: src/PowerShellEditorServices/Session/PowerShellContext.cs [hunk 10]
```diff
@@ -434,28 +486,57 @@ public Task<RunspaceHandle> GetRunspaceHandle(CancellationToken cancellationToke
...
             {
                 this.logger.Write(LogLevel.Verbose, "Passing command execution to pipeline thread.");
 
-                PipelineExecutionRequest<TResult> executionRequest =
+                if (shouldCancelReadLine && PromptNest.IsReadLineBusy())
+                {
+                    // If a ReadLine pipeline is running in the debugger then we'll hang here
+                    // if we don't cancel it. Typically we can rely on OnExecutionStatusChanged but
+                    // the pipeline request won't even start without clearing the current task.
+                    // await this.PromptContext.AbortReadLineAsync();
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/PowerShellContext.cs]: Remove commented code.

File: src/PowerShellEditorServices/Session/PowerShellContext.cs [hunk 11]
```diff
@@ -944,22 +1183,57 @@ internal void BreakExecution()
...
+                if (this.PromptNest.IsNestedPrompt)
+                {
+                    this.ExitAllNestedPrompts();
+                }
+
+                if (this.PromptNest.IsInDebugger)
+                {
+                    // Set the result so that the execution thread resumes.
+                    // The execution thread will clean up the task.
+
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/PowerShellContext.cs]: Extra new line.

File: src/PowerShellEditorServices/Session/PowerShellContext.cs [hunk 12]
```diff
@@ -1110,15 +1503,17 @@ internal void ReleaseRunspaceHandle(RunspaceHandle runspaceHandle)
...
-                runspaceHandle.Runspace.SessionStateProxy.Path.SetLocation(path);
+                path = EscapePath(path, false);
             }
+
+            await ExecuteCommand<PSObject>(
+                new PSCommand().AddCommand("Set-Location").AddParameter("Path", path),
+                null,
+                false,
+                false,
+                false);
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/PowerShellContext.cs]: `bool` parameters should be named.

File: src/PowerShellEditorServices/Session/PowerShellContext.cs [hunk 13]
```diff
@@ -1239,7 +1634,7 @@ private IEnumerable<TResult> ExecuteCommandInDebugger<TResult>(PSCommand psComma
             if (debuggerResumeAction.HasValue)
             {
                 // Resume the debugger with the specificed action
-                this.ResumeDebugger(debuggerResumeAction.Value);
+                this.ResumeDebugger(debuggerResumeAction.Value, false);
```
Comment (SeeminglyScience) [inline src/PowerShellEditorServices/Session/PowerShellContext.cs]: `bool` parameter should be named.

File: src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs (hunk 1 above)
Comment (SeeminglyScience) [reply src/PowerShellEditorServices/Session/PSReadLinePromptContext.cs]: It does indeed :)

Mostly anyway.  The commands we invoke in the background still show up in `Get-History` and still increment `$MyInvocation.HistoryId` which is annoying.  But navigating history with all the PSRL hotkeys like CTRL + R/up/down/etc work as expected in the debugger as well :)

**Comment by SeeminglyScience** [comment]
Had some issues with merge conflict resolution, continuing with PR #672

**Pull request closed without merging**